From 9ed69eaed35f8d7b6032241a1e0874dc215c61f7 Mon Sep 17 00:00:00 2001
From: tobigun <hennymcc@yahoo.de>
Date: Thu, 31 Dec 2015 15:47:03 +0100
Subject: [PATCH] ril: reintroduce changes necessary to make SIM detection work

libril checks the size of RIL_CardStatus to detect whether it is v6 or
v5. As Samsung added some additional fields the version cannot be
detected correctly anymore.

Change-Id: I0d786787e0416d110b9930f9a72384d0f50dcd5d
---
 include/telephony/ril.h | 149 ++++++++++++++++++++++++++++++++++++++++++++++++
 rild/rild.c             |   9 ++-
 2 files changed, 156 insertions(+), 2 deletions(-)

diff --git a/include/telephony/ril.h b/include/telephony/ril.h
index 2e71367..09cc5a8 100644
--- a/include/telephony/ril.h
+++ b/include/telephony/ril.h
@@ -289,6 +289,12 @@ typedef struct {
 } RIL_CDMA_SignalInfoRecord;
 
 typedef struct {
+    char            call_type;  /* Samsung */
+    char            call_domain;/* Samsung */
+    char *          csv;        /* Samsung */
+} RIL_Call_extra;
+
+typedef struct {
     RIL_CallState   state;
     int             index;      /* Connection Index for use with, eg, AT+CHLD */
     int             toa;        /* type of address, eg 145 = intl */
@@ -297,12 +303,17 @@ typedef struct {
     char            als;        /* ALS line indicator if available
                                    (0 = line 1) */
     char            isVoice;    /* nonzero if this is is a voice call */
+    char            isVideo;    /* Samsung */
     char            isVoicePrivacy;     /* nonzero if CDMA voice privacy mode is active */
+    char            unknown1;
+    char            unknown2;
     char *          number;     /* Remote party number */
     int             numberPresentation; /* 0=Allowed, 1=Restricted, 2=Not Specified/Unknown 3=Payphone */
     char *          name;       /* Remote party name */
     int             namePresentation; /* 0=Allowed, 1=Restricted, 2=Not Specified/Unknown 3=Payphone */
+    int             unknown3;
     RIL_UUS_Info *  uusInfo;    /* NULL or Pointer to User-User Signaling Information */
+    RIL_Call_extra * extra;
 } RIL_Call;
 
 /* Deprecated, use RIL_Data_Call_Response_v6 */
@@ -503,6 +514,7 @@ typedef struct {
     char *data;     /* May be NULL*/
     char *pin2;     /* May be NULL*/
     char *aidPtr;   /* AID value, See ETSI 102.221 8.1 and 101.220 4, NULL if no value. */
+    int cla;	    /* Class of the APDU command */
 } RIL_SIM_IO_v6;
 
 /* Used by RIL_REQUEST_SIM_TRANSMIT_APDU_CHANNEL and
@@ -770,6 +782,12 @@ typedef struct
   int              pin1_replaced;   /* applicable to USIM, CSIM & ISIM */
   RIL_PinState     pin1;
   RIL_PinState     pin2;
+  /* Samsung enhancements */
+  int              pin1_num_retries;
+  int              puk1_num_retries;
+  int              pin2_num_retries;
+  int              puk2_num_retries;
+  int              perso_unblock_retries;
 } RIL_AppStatus;
 
 /* Deprecated, use RIL_CardStatus_v6 */
@@ -4462,6 +4480,70 @@ typedef struct {
 
 /***********************************************************************/
 
+/* SAMSUNG REQUESTS */
+#define RIL_REQUEST_SELECT_BAND 509
+#define RIL_REQUEST_GET_BAND 510
+#define RIL_REQUEST_SET_FDY 525
+#define RIL_REQUEST_SET_COMCFG 526
+#define RIL_REQUEST_GET_COMCFG 527
+#define RIL_REQUEST_SWITCH_MODEM 528
+
+#define RIL_REQUEST_GET_CELL_BROADCAST_CONFIG 10002
+
+#define RIL_REQUEST_SEND_ENCODED_USSD 10005
+#define RIL_REQUEST_SET_PDA_MEMORY_STATUS 10006
+#define RIL_REQUEST_GET_PHONEBOOK_STORAGE_INFO 10007
+#define RIL_REQUEST_GET_PHONEBOOK_ENTRY 10008
+#define RIL_REQUEST_ACCESS_PHONEBOOK_ENTRY 10009
+#define RIL_REQUEST_DIAL_VIDEO_CALL 10010
+#define RIL_REQUEST_CALL_DEFLECTION 10011
+//#define RIL_REQUEST_READ_SMS_FROM_SIM 10012 /*UNKNOWN*/
+#define RIL_REQUEST_USIM_PB_CAPA 10013
+#define RIL_REQUEST_LOCK_INFO 10014
+
+#define RIL_REQUEST_DIAL_EMERGENCY 10016
+//#define RIL_REQUEST_GET_STOREAD_MSG_COUNT 10017 /*UNKNOWN*/
+#define RIL_REQUEST_STK_SIM_INIT_EVENT 10018
+#define RIL_REQUEST_GET_LINE_ID 10019
+#define RIL_REQUEST_SET_LINE_ID 10020
+#define RIL_REQUEST_GET_SERIAL_NUMBER 10021
+#define RIL_REQUEST_GET_MANUFACTURE_DATE_NUMBER 10022
+#define RIL_REQUEST_GET_BARCODE_NUMBER 10023
+#define RIL_REQUEST_UICC_GBA_AUTHENTICATE_BOOTSTRAP 10024
+#define RIL_REQUEST_UICC_GBA_AUTHENTICATE_NAF 10025
+//#define RIL_REQUEST_SIM_AUTH 10030 /*UNKNOWN*/
+#define RIL_REQUEST_MODIFY_CALL_INITIATE 10031
+#define RIL_REQUEST_MODIFY_CALL_CONFIRM 10032
+#define RIL_REQUEST_SAFE_MODE 10033
+#define RIL_REQUEST_SET_VOICE_DOMAIN_PREF 10034
+#define RIL_REQUEST_PS_ATTACH 10035
+#define RIL_REQUEST_PS_DETACH 10036
+#define RIL_REQUEST_ACTIVATE_DATA_CALL 10037
+#define RIL_REQUEST_CHANGE_SIM_PERSO 10038
+#define RIL_REQUEST_ENTER_SIM_PERSO 10039
+#define RIL_REQUEST_GET_TIME_INFO 10040
+#define RIL_REQUEST_CDMA_SEND_SMS_EXPECT_MORE 10041
+#define RIL_REQUEST_OMADM_SETUP_SESSION 10042
+#define RIL_REQUEST_OMADM_SERVER_START_SESSION 10043
+#define RIL_REQUEST_OMADM_CLIENT_START_SESSION 10044
+#define RIL_REQUEST_OMADM_SEND_DATA 10045
+#define RIL_REQUEST_CDMA_GET_DATAPROFILE 10046
+#define RIL_REQUEST_CDMA_SET_DATAPROFILE 10047
+#define RIL_REQUEST_CDMA_GET_SYSTEMPROPERTIES 10048
+#define RIL_REQUEST_CDMA_SET_SYSTEMPROPERTIES 10049
+#define RIL_REQUEST_SEND_SMS_COUNT 10050
+#define RIL_REQUEST_SEND_SMS_MSG 10051
+#define RIL_REQUEST_SEND_SMS_MSG_READ_STATUS 10052
+#define RIL_REQUEST_MODEM_HANGUP 10053
+#define RIL_REQUEST_SET_SIM_POWER 10054
+#define RIL_REQUEST_SET_PREFERRED_NETWORK_LIST 10055
+#define RIL_REQUEST_GET_PREFERRED_NETWORK_LIST 10056
+#define RIL_REQUEST_HANGUP_VT 10057
+#define RIL_REQUEST_HOLD 10058
+#define RIL_REQUEST_SET_LTE_BAND_MODE 10059
+#define RIL_REQUEST_QUERY_LOCK_NETWORKS 10060
+
+/***********************************************************************/
 
 #define RIL_UNSOL_RESPONSE_BASE 1000
 
@@ -5061,6 +5143,73 @@ typedef struct {
 
 /***********************************************************************/
 
+#define SAMSUNG_UNSOL_RESPONSE_BASE 11000
+
+#define RIL_UNSOL_RELEASE_COMPLETE_MESSAGE 11001
+#define RIL_UNSOL_STK_SEND_SMS_RESULT 11002
+#define RIL_UNSOL_STK_CALL_CONTROL_RESULT 11003
+#define RIL_UNSOL_DUN_CALL_STATUS 11004
+
+#define RIL_UNSOL_O2_HOME_ZONE_INFO 11007
+#define RIL_UNSOL_DEVICE_READY_NOTI 11008
+#define RIL_UNSOL_GPS_NOTI 11009
+#define RIL_UNSOL_AM 11010
+#define RIL_UNSOL_DUN_PIN_CONTROL_SIGNAL 11011
+#define RIL_UNSOL_DATA_SUSPEND_RESUME 11012
+#define RIL_UNSOL_SAP 11013
+
+//#define RIL_UNSOL_SIM_SMS_STORAGE_AVAILALE 11015 /*UNKNOWN*/
+#define RIL_UNSOL_HSDPA_STATE_CHANGED 11016
+#define RIL_UNSOL_WB_AMR_STATE 11017
+#define RIL_UNSOL_TWO_MIC_STATE 11018
+#define RIL_UNSOL_DHA_STATE 11019
+#define RIL_UNSOL_UART 11020
+#define RIL_UNSOL_SIM_PB_READY 11021
+#define RIL_UNSOL_PCMCLOCK_STATE 11022
+#define RIL_UNSOL_1X_SMSPP 11023
+
+#define RIL_UNSOL_IMS_REGISTRATION_STATE_CHANGED 11027
+#define RIL_UNSOL_MODIFY_CALL 11028
+#define RIL_UNSOL_SRVCC_HANDOVER 11029
+#define RIL_UNSOL_CS_FALLBACK 11030
+
+/**
+ * RIL_UNSOL_UICC_SUBSCRIPTION_STATUS_CHANGED
+ *
+ * Indicated when there is a change in subscription status.
+ * This event will be sent in the following scenarios
+ *  - subscription readiness at modem, which was selected by telephony layer
+ *  - when subscription is deactivated by modem due to UICC card removal
+ *  - When network invalidates the subscription i.e. attach reject due to authentication reject
+ *
+ * "data" is const int *
+ * ((const int *)data)[0] == 0 for Subscription Deactivated
+ * ((const int *)data)[0] == 1 for Subscription Activated
+ *
+ */
+#define RIL_UNSOL_UICC_SUBSCRIPTION_STATUS_CHANGED 11031
+
+#define UNSOL_VOICE_SYSTEM_ID 11032
+
+#define RIL_UNSOL_IMS_RETRYOVER 11034
+#define RIL_UNSOL_STK_ALPHA_ID 11035
+#define RIL_UNSOL_PB_INIT_COMPLETE 11036
+
+#define RIL_UNSOL_RTS_INDICATION 11041
+#define RIL_UNSOL_RILD_RESET_NOTI 11042
+#define RIL_UNSOL_HOME_NETWORK_NOTI 11043
+
+#define RIL_UNSOL_OMADM_SEND_DATA 11045
+#define RIL_UNSOL_DUN 11046
+#define RIL_UNSOL_SYSTEM_REBOOT 11047
+#define RIL_UNSOL_VOICE_PRIVACY_CHANGED 11048
+#define RIL_UNSOL_UTS_GETSMSCOUNT 11049
+#define RIL_UNSOL_UTS_GETSMSMSG 11050
+#define RIL_UNSOL_UTS_GET_UNREAD_SMS_STATUS 11051
+#define RIL_UNSOL_MIP_CONNECT_STATUS 11052
+#define RIL_UNSOL_STK_CALL_STATUS 11053
+
+/*************************************************************************/
 
 #if defined(ANDROID_MULTI_SIM)
 /**
diff --git a/rild/rild.c b/rild/rild.c
index efc7c9d..3b9d9b1 100644
--- a/rild/rild.c
+++ b/rild/rild.c
@@ -68,6 +68,8 @@ extern void RIL_onUnsolicitedResponse(int unsolResponse, const void *data,
         size_t datalen);
 #endif
 
+extern int RIL_getCpCrashed();
+
 extern void RIL_requestTimedCallback (RIL_TimedCallback callback,
         void *param, const struct timeval *relativeTime);
 
@@ -368,7 +370,10 @@ OpenLib:
 done:
 
     RLOGD("RIL_Init starting sleep loop");
-    while (true) {
-        sleep(UINT32_MAX);
+    while (!RIL_getCpCrashed()) {
+        sleep(1);
     }
+
+    RLOGD("CP crashed -> restart");
+    return 1;
 }
-- 
1.9.1

